### DISCUSSION TITLE
Convergence issue in phase-field (KKS model)

### DISCUSSION THREAD

**Discussion by GuoChi-Li** [idaholab/moose#22386] [Q&A Modules: Phase field] [answered]
Hello MOOSE developers,

I want to simulate a KKS solidification model in MOOSE using the phase-field model, so I find a paper for this purpose. The paper's title is '_Phase-field simulation of structure evolution at high growth velocities during directional solidification of Ti55Al45 alloys_', and the DOI is https://doi.org/10.1016/j.intermet.2004.07.018. I think it is an excellent paper to start because it combines the CALPHAD (CALculation of PHAse Diagrams) data into a phase field method. However, I have encountered a convergence issue. The output message was 'Nonlinear solve did not converge due to DIVERGED_FNORM_NAN iterations 0.' I am not sure why this happened. 

These are my input file and output file.
[kks_model_setup_Ti_Al.log](https://github.com/idaholab/moose/files/9772367/kks_model_setup_Ti_Al.log)
[output_LU.log](https://github.com/idaholab/moose/files/9772359/output_LU.log)

I also try different solver conditions, such as LU, BoomerAMG, and ASM with the PETSc conditions in the input file.

And the second question is that the paper implements $[(1-c)W_A+cW_B ]*g(\phi)$ as shown in the first term in equation 4. I think the different W is not supported in moose. Correct me if I am wrong. Is there any chance I can implement this? Thank you for your time.

Sincerely,

Guo-Chi

**Comment by GiudGiud** [comment, marked as answer]
Hello

The solver does not solve any time step, it sees a NaN and cant proceed.
The culprit is likely one of these parsed expressions:
```
      function = '(c*GAl_L+(1-c)*GTi_L+R*T*(c*log(c)+(1-c)*log(1-c))+c*(1-c)*(G0_L+G1_L*(1-2*c)))/Vm'
```
if c becomes negative or null or above 1, this will generate a NaN. I would recommend either:
- adding bounds in the function to avoid nans at all times (max(0, min(c, 1)) for example
- using the bounds system on c
https://mooseframework.inl.gov/source/auxkernels/ConstantBoundsAux.html

Guillaume

**Reply by GiudGiud** [reply]
Update on this:
I've tried --trap-fpe and running in Debug mode but I cant get the solver to catch what is causing the NaN.

**Reply by GuoChi-Li** [reply]
Hi GiudGiud,

Thank for your time and your explanation. I am not sure how to add bounds in the function, but I have tried using the ‘bounds system’ on c. However, it still gets NaN. I also tried adding the bounds system on c, cl, cs, but this didn’t work.

I have two questions for the bounds system. (1)What is the function of the ‘ bounds_dummy’? Is this just a AuxVariables for the bounds systems or it has other physical meaning? (2) After I added the bounds system, MOOSE told me I should use ‘a variational inequalities solver.’ Does the ‘LU’ or the other solver (such as BoomerAMG or ASM ) support this type of problem?

Sincerely,

Guo-Chi

**Reply by hugary1995** [reply]
1. bounds_dummy, as its name suggests, is a dummy variable helping the system figure out which dofs the constraint applies to.
2. Popular petsc VI solvers are [VNEWTONRSLS](https://petsc.org/main/docs/manualpages/SNES/SNESVINEWTONRSLS/) and [VINEWTONSSLS](https://petsc.org/main/docs/manualpages/SNES/SNESVINEWTONSSLS/). If you don't use a VI snes solver, then the bounds will be ignored. AFAIK, all linear preconditioners, e.g. lu, asm, amg, are compatible with snes VI solvers.

**Reply by GuoChi-Li** [reply]
Hi hugary1995,
Thank you for your explanations. I have added
```
[Bounds]
  [./c_upper_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c
    bound_type = upper
    bound_value = 1
  [../]

  [./c_lower_bound]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c
    bound_type = lower
    bound_value = 0
  [../]
[]
```
and I have also tried **LU with vinewtonssls**, **LU with vinewtonrsls** in my code. The codes are shown as below.
```
# LU with vinewtonssls
  petsc_options_iname = '-pc_type -snes_type    -snes_vi_monitor'
  petsc_options_value = 'lu       vinewtonssls  true'

# LU with vinewtonrsls
  petsc_options_iname = '-pc_type -snes_type    -snes_vi_monitor'
  petsc_options_value = 'lu       vinewtonrsls  true'
```
However, I still got NaN in MOOSE.

**Reply by GiudGiud** [reply]
Hello

The NaNs must be coming from somewhere else then.

With LU that can happen with 0s on the diagonals.
Can you use a pc_factor_shift of NONZERO in your preconditioner? (goes in the petsc options)

Another thing would be to drastically reduce the system size to a few elements and print the matrix to check it with ksp_view_mat

Guillaume

**Reply by GuoChi-Li** [reply]
Hello GiudGiud,
Based on your advice, I have tried these two changes.
(1) I have removed the **petsc_options_iname**, **petsc_options_value** in the Executioner section, and changed the Preconditioning section as follows
```
[Preconditioning]
  [./full]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -snes_type    -snes_vi_monitor  -pc_factor_shift_type'
    petsc_options_value = 'lu       vinewtonssls  true              NONZERO'
  [../]
[]
```
I still got NaNs.
I also added -pc_factor_shift_amount 1e-9 in the petsc options, but this does not help.

(2) I have reduced the mesh from 900 * 1000 to 10*10 (also modified the initial condition slightly) and changed the Preconditioning by adding the **-ksp_view_mat binary** as the petsc document mentioned. (https://petsc.org/release/docs/manualpages/KSP/KSPSolve/)
```
[Preconditioning]
  [./full]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -snes_type    -snes_vi_monitor  -pc_factor_shift_type -pc_factor_shift_amount -ksp_view_mat'
    petsc_options_value = 'lu       vinewtonssls  true              NONZERO               1e-9                    binary'
  [../]
[]
```
I got two files named **binaryoutput** and **binaryoutput.info**. But there is nothing in these files. Not sure if I am doing this correctly or not. I have also added **-ksp_view_mat binary** in the original input file (900 * 1000 mesh), but I still got the empty **binaryoutput** and **binaryoutput.info**.

Guo-Chi

**Reply by GiudGiud** [reply]
If you export it in binary you ll need to find a reader to look at the matrix. I think matlab can do it

**Reply by GuoChi-Li** [reply]
Can I export the file into other formats? (I only see **binary** on the website.) The file size is 0 bytes, so I don't think I did it correctly. Here are the two files. I have added **.txt** in the file end to make files acceptable by GitHub.
[binaryoutput.info.txt](https://github.com/idaholab/moose/files/9852694/binaryoutput.info.txt)
[binaryoutput.txt](https://github.com/idaholab/moose/files/9852695/binaryoutput.txt)

**Reply by GiudGiud** [reply]
I think if you pass it in petsc_options directly you get text

**Reply by GuoChi-Li** [reply]
I used Matlab to open the file with **fread** or **fopen**, but it is empty. I am unsure what you mean by 'pass it in petsc_options directly.' If I don't provide the same length of arguments in both **petsc_options_iname** and **petsc_options_value**, moose will get an error. Also, I have tried **-ksp_view_mat text**, but it is not supported in petsc.

**Reply by GiudGiud** [reply]
There is also
`petsc_options = '-ksp_view_mat'` in the [Executioner] block to pass a single petsc option

**Reply by GuoChi-Li** [reply]
OK. Now I understand. I have added `petsc_options = '-ksp_view_mat'` in the [Executioner] block. However, I don't see any file generated in the folder. Here is my output file.
[20221025_running_with_ksp_view_mat.log](https://github.com/idaholab/moose/files/9856699/20221025_running_with_ksp_view_mat.log)

**Reply by GuoChi-Li** [reply]
I may have solved the problem. It was due to the inaccurate initial conditions in the input file.
But, I encountered a new issue. When I ran the problem with one processor, it was ok. However, when I tried to use MPI, moose raised an error. It said **PC failed due to FACTOR_OUTMEMORY**. I have checked the memory in the system with different PETSc options. Sometimes, it only uses about half of the memory, but it still has this error. What might happen?

I have tried ilu as this [discussion](https://github.com/idaholab/moose/discussions/16288) advised.

**Reply by GiudGiud** [reply]
FACTOR_OUTMEMORY can be a catch all for several other issues.
Did you try other preconditioning methods?

A typical issue in phase field is material properties being evaluated in non physical configurations, like having a log of something negative. Can you paste your input file here?

**Reply by GuoChi-Li** [reply]
Sorry for the late reply. Yes, I have tried several preconditioning methods, but they didn't work. Only LU works.
This is my input file.
[kks_model_Ti_Al_rectangular.txt](https://github.com/idaholab/moose/files/10099845/kks_model_Ti_Al_rectangular.txt)
